From bioformats issue #4377, EOFException when initializing Imaris .ims file

### Versions impacted by the bug


maint-5.x


### What went wrong?

Just wanted to mention a possible code change and see if it’s thought to be needed.

This is in regards to Bioformats Issue  4377( https://github.com/ome/bioformats/issues/4377)  opened by  @melissalinkert,   That issue has the stack traces and screen recordings.

In using Bioformats bftools **showinf**, it gets an Exception when reading this ims file https://zenodo.org/records/17406102

From the stack trace it looks like the files involved are RandomAccessFile.java and FractalHeap.java 

I was able to reproduce the problem in Eclipse and did some looking around.
I see in FractalHeap.readIndirectBlock() in the first line(line 453),  the file pointer of the Random Access File gets changed.  Member variable **raf**.



https://github.com/Unidata/netcdf-java/blob/1ff3f0c9571a23b796dcb577ab1680fff28907ab/cdm/core/src/main/java/ucar/nc2/iosp/hdf5/FractalHeap.java#L453


  Then at lines 517 in  method FractalHeap.readIndirectBlock() there’s a recursive call to itself.
 
https://github.com/Unidata/netcdf-java/blob/1ff3f0c9571a23b796dcb577ab1680fff28907ab/cdm/core/src/main/java/ucar/nc2/iosp/hdf5/FractalHeap.java#L517

I thought what might be needed is to save the file pointer before the recursive call and restore it afterwards.

          long savePos = raf.getFilePointer();
          readIndirectBlock(iblock2, childIndirectAddress, heapAddress, hasFilter);
          raf.seek(savePos);
 

After making this code change in Eclipse the Exception went away and it looked like the entire file was read.  A window was displayed but I didn’t see any image data at all. There is a screen recording and a log file in https://github.com/ome/bioformats/issues/4377.



### Relevant stack trace

```shell
$ showinf -no-upgrade tmp.ims 
Checking file format [Bitplane Imaris 5.5 (HDF)]
Initializing reader
ImarisHDFReader initializing tmp.ims
Exception in thread "main" java.io.IOException: java.io.EOFException: Reading tmp.ims at 4776428349165719766 file length = 4776428349165719766
	at ucar.nc2.NetcdfFiles.open(NetcdfFiles.java:281)
	at ucar.nc2.NetcdfFiles.open(NetcdfFiles.java:245)
	at ucar.nc2.NetcdfFiles.open(NetcdfFiles.java:218)
	at loci.formats.services.NetCDFServiceImpl.init(NetCDFServiceImpl.java:311)
	at loci.formats.services.NetCDFServiceImpl.setFile(NetCDFServiceImpl.java:104)
	at loci.formats.in.ImarisHDFReader.initFile(ImarisHDFReader.java:260)
	at loci.formats.FormatReader.setId(FormatReader.java:1480)
	at loci.formats.ImageReader.setId(ImageReader.java:864)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:692)
	at loci.formats.tools.ImageInfo.testRead(ImageInfo.java:1054)
	at loci.formats.tools.ImageInfo.main(ImageInfo.java:1165)
Caused by: java.io.EOFException: Reading tmp.ims at 4776428349165719766 file length = 4776428349165719766
	at ucar.unidata.io.RandomAccessFile.readFully(RandomAccessFile.java:840)
	at ucar.unidata.io.RandomAccessFile.readFully(RandomAccessFile.java:819)
	at ucar.unidata.io.RandomAccessFile.readString(RandomAccessFile.java:1564)
	at ucar.unidata.io.RandomAccessFile.readString(RandomAccessFile.java:1550)
	at ucar.nc2.iosp.hdf5.FractalHeap.readIndirectBlock(FractalHeap.java:456)
	at ucar.nc2.iosp.hdf5.FractalHeap.readIndirectBlock(FractalHeap.java:517)
	at ucar.nc2.iosp.hdf5.FractalHeap.<init>(FractalHeap.java:192)
	at ucar.nc2.internal.iosp.hdf5.H5objects$DataObject.processAttributeInfoMessage(H5objects.java:423)
	at ucar.nc2.internal.iosp.hdf5.H5objects$DataObject.<init>(H5objects.java:407)
	at ucar.nc2.internal.iosp.hdf5.H5objects$DataObject.<init>(H5objects.java:238)
	at ucar.nc2.internal.iosp.hdf5.H5objects.readDataObject(H5objects.java:235)
	at ucar.nc2.internal.iosp.hdf5.H5headerNew.getDataObject(H5headerNew.java:2265)
	at ucar.nc2.internal.iosp.hdf5.H5objects$DataObjectFacade.<init>(H5objects.java:108)
	at ucar.nc2.internal.iosp.hdf5.H5objects.readGroupNew(H5objects.java:2154)
	at ucar.nc2.internal.iosp.hdf5.H5objects.access$700(H5objects.java:27)
	at ucar.nc2.internal.iosp.hdf5.H5objects$H5Group.<init>(H5objects.java:200)
	at ucar.nc2.internal.iosp.hdf5.H5objects$H5Group.<init>(H5objects.java:158)
	at ucar.nc2.internal.iosp.hdf5.H5objects.readH5Group(H5objects.java:155)
	at ucar.nc2.internal.iosp.hdf5.H5headerNew.makeNetcdfGroup(H5headerNew.java:524)
	at ucar.nc2.internal.iosp.hdf5.H5headerNew.read(H5headerNew.java:254)
	at ucar.nc2.internal.iosp.hdf5.H5iospNew.build(H5iospNew.java:126)
	at ucar.nc2.NetcdfFiles.build(NetcdfFiles.java:835)
	at ucar.nc2.NetcdfFiles.open(NetcdfFiles.java:770)
	at ucar.nc2.NetcdfFiles.open(NetcdfFiles.java:278)
	... 10 more
```

### Relevant log messages

_No response_

### If you have an example file that you can share, please attach it to this issue.
If so, may we include it in our test datasets to help ensure the bug does not return once fixed?
Note: the test datasets are publicly accessible without restriction.


Yes

### Code of Conduct

- [x] I agree to follow the UCAR/Unidata Code of Conduct

--- a/cdm-test/src/test/java/ucar/nc2/iosp/hdf5/TestH5IndirectBlocks.java
+++ b/cdm-test/src/test/java/ucar/nc2/iosp/hdf5/TestH5IndirectBlocks.java
@@ -1,0 +1,68 @@
+/*
+ * Copyright (c) 2026 University Corporation for Atmospheric Research/Unidata
+ * See LICENSE for license information.
+ */
+
+package ucar.nc2.iosp.hdf5;
+
+import static com.google.common.truth.Truth.assertThat;
+import static ucar.unidata.util.test.TestDir.cdmUnitTestDir;
+
+import java.io.IOException;
+import java.util.List;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import ucar.nc2.Group;
+import ucar.nc2.NetcdfFile;
+import ucar.nc2.NetcdfFiles;
+import ucar.nc2.Variable;
+import ucar.unidata.util.test.category.NeedsCdmUnitTest;
+
+@Category(NeedsCdmUnitTest.class)
+public class TestH5IndirectBlocks {
+
+  private final String h5FileWithIndirectBlocks = cdmUnitTestDir + "formats/hdf5/bitplane_imaris/indirect_blocks.ims";
+
+  // demonstrated https://github.com/Unidata/netcdf-java/issues/1546 using old api
+  @Test
+  public void testOpenFileWithIndirectBlocksOldApi() throws IOException {
+    NetcdfFile ncf = NetcdfFile.open(h5FileWithIndirectBlocks);
+    assertThat(ncf).isNotNull();
+  }
+
+  // demonstrated https://github.com/Unidata/netcdf-java/issues/1546 using new api
+  @Test
+  public void testOpenFileWithIndirectBlocksNewApi() throws IOException {
+    NetcdfFile ncf = NetcdfFiles.open(h5FileWithIndirectBlocks);
+    assertThat(ncf).isNotNull();
+  }
+
+  @Test
+  public void testReadIndirectBlocksOldApi() throws IOException {
+    NetcdfFile ncf = NetcdfFile.open(h5FileWithIndirectBlocks);
+    assertThat(ncf).isNotNull();
+    readVariableWithIndirectBlocks(ncf);
+  }
+
+  @Test
+  public void testReadIndirectBlocksNewApi() throws IOException {
+    NetcdfFile ncf = NetcdfFiles.open(h5FileWithIndirectBlocks);
+    assertThat(ncf).isNotNull();
+    readVariableWithIndirectBlocks(ncf);
+  }
+
+  private void readVariableWithIndirectBlocks(NetcdfFile ncf) throws IOException {
+    Group groupWithIndirectBlocksVariables = ncf.findGroup("/DataSetInfo/ZeissAttrs");
+    assertThat(groupWithIndirectBlocksVariables).isNotNull();
+    List<Variable> variablesWithIndirectBlocks = groupWithIndirectBlocksVariables.getVariables();
+    assertThat(variablesWithIndirectBlocks).isNotNull();
+    assertThat(variablesWithIndirectBlocks.size()).isEqualTo(3);
+    for (Variable variable : variablesWithIndirectBlocks) {
+      String data = variable.readScalarString();
+      assertThat(data).isNotNull();
+      assertThat(data).isNotEmpty();
+      assertThat(data).startsWith("<?xml");
+      assertThat(data).contains("<ArrayOfRt");
+    }
+  }
+}

--- a/cdm/core/src/main/java/ucar/nc2/iosp/hdf5/FractalHeap.java
+++ b/cdm/core/src/main/java/ucar/nc2/iosp/hdf5/FractalHeap.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1998-2018 John Caron and University Corporation for Atmospheric Research/Unidata
+ * Copyright (c) 1998-2026 John Caron and University Corporation for Atmospheric Research/Unidata
  * See LICENSE for license information.
  */
 
@@ -513,8 +513,11 @@ public class FractalHeap {
         long childIndirectAddress = h5.readOffset();
         if (debugDetail || debugFractalHeap)
           debugOut.println("  InDirectChild " + row + " address= " + childIndirectAddress);
-        if (childIndirectAddress >= 0)
+        if (childIndirectAddress >= 0) {
+          long savePos = raf.getFilePointer();
           readIndirectBlock(iblock2, childIndirectAddress, heapAddress, hasFilter);
+          raf.seek(savePos);
+        }
       }
     }
 

